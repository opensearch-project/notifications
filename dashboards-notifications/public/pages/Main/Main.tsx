/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

/*
 * Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import { EuiPage, EuiPageBody, EuiPageSideBar, EuiSideNav } from '@elastic/eui';
import React, { Component, createContext } from 'react';
import { Redirect, Route, RouteComponentProps, Switch } from 'react-router-dom';
import { CoreStart } from '../../../../../src/core/public';
import { CoreServicesConsumer } from '../../components/coreServices';
import { ModalProvider, ModalRoot } from '../../components/Modal';
import { BrowserServices } from '../../models/interfaces';
import { ServicesConsumer, ServicesContext } from '../../services/services';
import { CHANNEL_TYPE, ROUTES } from '../../utils/constants';
import { Channels } from '../Channels/Channels';
import { ChannelDetails } from '../Channels/components/details/ChannelDetails';
import { CreateChannel } from '../CreateChannel/CreateChannel';
import { CreateRecipientGroup } from '../Emails/CreateRecipientGroup';
import { CreateSender } from '../Emails/CreateSender';
import { EmailGroups } from '../Emails/EmailGroups';
import Notifications from '../Notifications';

enum Navigation {
  Notifications = 'Notifications',
  Dashboard = 'Dashboard',
  Channels = 'Channels',
  EmailGroups = 'Email groups',
}

enum Pathname {
  Notifications = '/notifications',
  Channels = '/channels',
}

interface MainProps extends RouteComponentProps {}

export interface MainState {
  availableFeatures: Partial<typeof CHANNEL_TYPE>;
  tooltipSupport: boolean; // if true, IAM role for SNS is optional and helper text should be available
}

export const MainContext = createContext<MainState | null>(null);

export default class Main extends Component<MainProps, MainState> {
  static contextType = ServicesContext;

  constructor(props: MainProps) {
    super(props);
    this.state = {
      availableFeatures: CHANNEL_TYPE,
      tooltipSupport: false,
    };
  }

  async componentDidMount() {
    const serverFeatures = await this.context.notificationService.getServerFeatures();
    if (serverFeatures != null)
      this.setState({
        availableFeatures: serverFeatures.availableFeatures,
        tooltipSupport: serverFeatures.tooltipSupport,
      });
  }

  render() {
    const {
      location: { pathname },
    } = this.props;

    const sideNav = [
      {
        name: Navigation.Notifications,
        id: 0,
        href: `#${Pathname.Notifications}`,
        items: [
          {
            name: Navigation.Dashboard,
            id: 1,
            href: `#${Pathname.Notifications}`,
            isSelected: pathname === Pathname.Notifications,
          },
          {
            name: Navigation.Channels,
            id: 2,
            href: `#${Pathname.Channels}`,
            isSelected: pathname === Pathname.Channels,
          },
          {
            name: Navigation.EmailGroups,
            id: 3,
            href: `#${ROUTES.EMAIL_GROUPS}`,
            isSelected: pathname === ROUTES.EMAIL_GROUPS,
          },
        ],
      },
    ];
    return (
      <CoreServicesConsumer>
        {(core: CoreStart | null) =>
          core && (
            <ServicesConsumer>
              {(services: BrowserServices | null) =>
                services && (
                  <MainContext.Provider value={this.state}>
                    <ModalProvider>
                      <ModalRoot services={services} />
                      <EuiPage>
                        {/*Hide side navigation bar when creating or editing rollup job*/}
                        {pathname !== ROUTES.CREATE_CHANNEL &&
                          !pathname.startsWith(ROUTES.EDIT_CHANNEL) &&
                          !pathname.startsWith(ROUTES.CHANNEL_DETAILS) &&
                          pathname !== ROUTES.CREATE_SENDER &&
                          !pathname.startsWith(ROUTES.EDIT_SENDER) &&
                          pathname !== ROUTES.CREATE_RECIPIENT_GROUP &&
                          !pathname.startsWith(ROUTES.EDIT_RECIPIENT_GROUP) && (
                            <EuiPageSideBar style={{ minWidth: 150 }}>
                              <EuiSideNav
                                style={{ width: 150 }}
                                items={sideNav}
                              />
                            </EuiPageSideBar>
                          )}
                        <EuiPageBody>
                          <Switch>
                            <Route
                              path={ROUTES.CREATE_CHANNEL}
                              render={(props: RouteComponentProps) => (
                                <CreateChannel {...props} />
                              )}
                            />
                            <Route
                              path={`${ROUTES.EDIT_CHANNEL}/:id`}
                              render={(
                                props: RouteComponentProps<{ id: string }>
                              ) => <CreateChannel {...props} edit={true} />}
                            />
                            <Route
                              path={`${ROUTES.CHANNEL_DETAILS}/:id`}
                              render={(
                                props: RouteComponentProps<{ id: string }>
                              ) => <ChannelDetails {...props} />}
                            />
                            <Route
                              path={ROUTES.CHANNELS}
                              render={(props: RouteComponentProps) => (
                                <Channels
                                  {...props}
                                  notificationService={
                                    services.notificationService
                                  }
                                />
                              )}
                            />
                            <Route
                              path={ROUTES.NOTIFICATIONS}
                              render={(props: RouteComponentProps) => (
                                <Notifications
                                  {...props}
                                  services={services}
                                  mainProps={this.state}
                                />
                              )}
                            />
                            <Route
                              path={ROUTES.EMAIL_GROUPS}
                              render={(props: RouteComponentProps) => (
                                <EmailGroups {...props} />
                              )}
                            />
                            <Route
                              path={ROUTES.CREATE_SENDER}
                              render={(props: RouteComponentProps) => (
                                <CreateSender {...props} />
                              )}
                            />
                            <Route
                              path={`${ROUTES.EDIT_SENDER}/:id`}
                              render={(props: RouteComponentProps) => (
                                <CreateSender {...props} edit={true} />
                              )}
                            />
                            <Route
                              path={ROUTES.CREATE_RECIPIENT_GROUP}
                              render={(props: RouteComponentProps) => (
                                <CreateRecipientGroup {...props} />
                              )}
                            />
                            <Route
                              path={`${ROUTES.EDIT_RECIPIENT_GROUP}/:id`}
                              render={(props: RouteComponentProps) => (
                                <CreateRecipientGroup {...props} edit={true} />
                              )}
                            />
                            <Redirect from="/" to={ROUTES.NOTIFICATIONS} />
                          </Switch>
                        </EuiPageBody>
                      </EuiPage>
                    </ModalProvider>
                  </MainContext.Provider>
                )
              }
            </ServicesConsumer>
          )
        }
      </CoreServicesConsumer>
    );
  }
}
