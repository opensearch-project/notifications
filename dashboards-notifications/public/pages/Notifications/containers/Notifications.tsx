/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

/*
 * Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import { Datum } from '@elastic/charts';
import dateMath from '@elastic/datemath';
import {
  EuiButton,
  EuiFlexGroup,
  EuiFlexItem,
  EuiSpacer,
  EuiTableSortingType,
  EuiTitle,
  ShortDate,
  SortDirection,
} from '@elastic/eui';
import { Criteria } from '@elastic/eui/src/components/basic_table/basic_table';
import { Pagination } from '@elastic/eui/src/components/basic_table/pagination_bar';
import _ from 'lodash';
import queryString from 'querystring';
import React, { Component } from 'react';
import { RouteComponentProps } from 'react-router-dom';
import { NotificationItem, TableState } from '../../../../models/interfaces';
import { CoreServicesContext } from '../../../components/coreServices';
import { BrowserServices } from '../../../models/interfaces';
import {
  BREADCRUMBS,
  CHANNEL_TYPE,
  HISTOGRAM_TYPE,
  ROUTES,
} from '../../../utils/constants';
import { getErrorMessage } from '../../../utils/helpers';
import { MainState } from '../../Main/Main';
import { EmptyState } from '../components/EmptyState/EmptyState';
import { NotificationsTable } from '../components/NotificationsTable/NotificationsTable';
import { FilterType } from '../components/SearchBar/Filter/Filters';
import { NotificationsSearchBar } from '../components/SearchBar/NotificationsSearchBar';
import { filtersToQueryParams } from '../components/SearchBar/utils/filterHelpers';
import { DEFAULT_PAGE_SIZE_OPTIONS } from '../utils/constants';
import { getURLQueryParams } from '../utils/helpers';

interface NotificationsProps extends RouteComponentProps {
  services: BrowserServices;
  mainProps: MainState;
}

interface NotificationsState extends TableState<NotificationItem> {
  startTime: ShortDate;
  endTime: ShortDate;
  filters: Array<FilterType>;
  histogramType: keyof typeof HISTOGRAM_TYPE;
  histogramData: Array<Datum>;
  isChannelConfigured: boolean;
}

export default class Notifications extends Component<
  NotificationsProps,
  NotificationsState
> {
  static contextType = CoreServicesContext;

  constructor(props: NotificationsProps) {
    super(props);

    const urlParams =
      this.props.location.search ||
      sessionStorage.getItem('NotificationsQueryParams') ||
      '';
    const {
      from,
      size,
      search,
      sortField,
      sortDirection,
      startTime,
      endTime,
      filters,
      histogramType,
    } = getURLQueryParams(urlParams);

    this.state = {
      total: 0,
      from,
      size,
      search,
      sortField,
      sortDirection,
      items: [],
      selectedItems: [],
      loading: true,
      startTime,
      endTime,
      filters,
      histogramType,
      histogramData: [],
      isChannelConfigured: true,
    };

    this.getNotifications = _.debounce(this.getNotifications, 200);
  }

  async componentDidMount() {
    this.context.chrome.setBreadcrumbs([
      BREADCRUMBS.NOTIFICATIONS,
      BREADCRUMBS.DASHBOARD,
    ]);
    window.scrollTo(0, 0);
    await this.getNotifications();
    try {
      const isChannelConfigured =
        await this.props.services.notificationService.getChannels({
          config_type: Object.keys(CHANNEL_TYPE),
          from_index: 0,
          max_items: 1,
          sort_field: 'name',
          sort_order: SortDirection.ASC,
        });
      if (!isChannelConfigured?.total) {
        this.setState({ isChannelConfigured: false });
      }
    } catch (error) {
      this.setState({ isChannelConfigured: false });
    }
  }

  async componentDidUpdate(
    prevProps: NotificationsProps,
    prevState: NotificationsState
  ) {
    const prevQuery = Notifications.serializeSearchParams(prevState);
    const currQuery = Notifications.serializeSearchParams(this.state);
    if (!_.isEqual(prevQuery, currQuery)) {
      await this.getNotifications();
    }
  }

  static serializeSearchParams(state: NotificationsState) {
    return {
      from_index: state.from,
      max_items: state.size,
      query: state.search,
      sort_field: state.sortField,
      sort_order: state.sortDirection,
      startTime: state.startTime,
      endTime: state.endTime,
      filters: JSON.stringify(state.filters),
      // histogramType: state.histogramType,
    };
  }

  static getQueryObjectFromState(state: NotificationsState) {
    const filterParams = filtersToQueryParams(state.filters);
    return {
      from_index: state.from,
      max_items: state.size,
      query: state.search,
      sort_field: state.sortField,
      sort_order: state.sortDirection,
      last_updated_time_ms: `${dateMath
        .parse(state.startTime)
        ?.valueOf()}..${dateMath.parse(state.endTime)?.valueOf()}`,
      ...filterParams,
    };
  }

  getNotifications = async () => {
    this.setState({ loading: true });
    try {
      const { services, history } = this.props;
      const searchParams = Notifications.serializeSearchParams(this.state);
      const queryParamsString = queryString.stringify(searchParams);
      history.replace({ ...this.props.location, search: queryParamsString });
      sessionStorage.setItem('NotificationsQueryParams', queryParamsString);

      const queryObject = Notifications.getQueryObjectFromState(this.state);
      const getNotificationsResponse = await services.eventService.getNotifications(
        queryObject
      );
      // const getHistogramResponse = await services.eventService.getHistogram(
      //   queryObject
      // );
      this.setState({
        items: getNotificationsResponse.items,
        total: getNotificationsResponse.total,
        // histogramData: getHistogramResponse,
      });
    } catch (err) {
      this.context.notifications.toasts.addDanger(
        getErrorMessage(err, 'There was a problem loading notifications.')
      );
    }
    this.setState({ loading: false });
  };

  onTableChange = ({
    page: tablePage,
    sort,
  }: Criteria<NotificationItem>): void => {
    const { index: page, size } = tablePage!;
    const { field: sortField, direction: sortDirection } = sort!;
    this.setState({ from: page * size, size, sortField, sortDirection });
  };

  onSearchChange = (search: string): void => {
    this.setState({ from: 0, search });
  };

  setStartTime = (startTime: ShortDate) => {
    this.setState({ from: 0, startTime });
  };
  setEndTime = (endTime: ShortDate) => {
    this.setState({ from: 0, endTime });
  };
  setFilters = (filters: FilterType[]) => {
    this.setState({ from: 0, filters });
    this.getNotifications();
  };
  setHistogramType = (histogramType: keyof typeof HISTOGRAM_TYPE) => {
    this.setState({ histogramType });
  };

  render() {
    const page = Math.floor(this.state.from / this.state.size);

    const pagination: Pagination = {
      pageIndex: page,
      pageSize: this.state.size,
      pageSizeOptions: DEFAULT_PAGE_SIZE_OPTIONS,
      totalItemCount: this.state.total,
    };

    const sorting: EuiTableSortingType<NotificationItem> = {
      sort: {
        direction: this.state.sortDirection,
        field: this.state.sortField,
      },
    };

    if (
      !this.state.loading &&
      this.state.total === 0 &&
      !this.state.isChannelConfigured
    ) {
      return <EmptyState channels={false} />;
    }

    return (
      <div style={{ padding: '0px 25px' }}>
        <EuiFlexGroup alignItems="center">
          <EuiFlexItem>
            <EuiTitle size="l">
              <h1>Dashboard</h1>
            </EuiTitle>
          </EuiFlexItem>
          <EuiFlexItem grow={false}>
            <EuiButton
              href={`#${ROUTES.CHANNELS}`}
              data-test-subj="createChannelButton"
              fill
            >
              Manage channels
            </EuiButton>
          </EuiFlexItem>
        </EuiFlexGroup>

        <EuiSpacer size="m" />
        <NotificationsSearchBar
          startTime={this.state.startTime}
          setStartTime={this.setStartTime}
          endTime={this.state.endTime}
          setEndTime={this.setEndTime}
          search={this.state.search}
          setSearch={this.onSearchChange}
          filters={this.state.filters}
          setFilters={this.setFilters}
          refresh={this.getNotifications}
        />

        {this.state.loading || this.state.total > 0 ? (
          <>
            {/* <EuiSpacer />
            <NotificationsHistogram
              histogramType={this.state.histogramType}
              setHistogramType={this.setHistogramType}
              histogramData={this.state.histogramData}
            /> */}

            <EuiSpacer />
            <NotificationsTable
              items={this.state.items}
              onTableChange={this.onTableChange}
              pagination={pagination}
              sorting={sorting}
              loading={this.state.loading}
            />
          </>
        ) : (
          <EmptyState channels={true} />
        )}
      </div>
    );
  }
}
