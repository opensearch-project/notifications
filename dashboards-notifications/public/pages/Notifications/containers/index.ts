/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import Notifications from './Notifications';

export default Notifications;
