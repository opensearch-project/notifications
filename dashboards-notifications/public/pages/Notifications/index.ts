/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import Notifications from './containers/Notifications';

export default Notifications;
