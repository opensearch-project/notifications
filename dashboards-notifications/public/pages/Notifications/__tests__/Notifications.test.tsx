/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import { fireEvent, render } from '@testing-library/react';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';
import { RouteComponentProps } from 'react-router-dom';
import Notifications from '..';
import {
  coreServicesMock,
  mainStateMock,
  notificationServiceMock,
} from '../../../../test/mocks/serviceMock';
import { CoreServicesContext } from '../../../components/coreServices';
import { ServicesContext } from '../../../services';
import { MainContext } from '../../Main/Main';

describe('<Notifications /> spec', () => {
  configure({ adapter: new Adapter() });

  it('renders the component', () => {
    const props = {
      location: { search: '' },
      match: { params: { id: 'test' } },
    };
    const utils = render(
      <MainContext.Provider value={mainStateMock}>
        <ServicesContext.Provider value={notificationServiceMock}>
          <CoreServicesContext.Provider value={coreServicesMock}>
            <Notifications
              {...(props as RouteComponentProps<{ id: string }>)}
              services={notificationServiceMock}
              mainProps={mainStateMock}
            />
          </CoreServicesContext.Provider>
        </ServicesContext.Provider>
      </MainContext.Provider>
    );

    fireEvent.change(utils.getByTestId('notifications-search-bar-input'), {
      target: { value: 'test' },
    });
    utils.getByText('Refresh').click();
    expect(utils.container.firstChild).toMatchSnapshot();
  });
});
