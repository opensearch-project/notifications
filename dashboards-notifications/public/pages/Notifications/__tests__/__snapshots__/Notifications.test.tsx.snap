// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Notifications /> spec renders the component 1`] = `
<div
  style="padding: 0px 25px;"
>
  <div
    class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--directionRow euiFlexGroup--responsive"
  >
    <div
      class="euiFlexItem"
    >
      <h1
        class="euiTitle euiTitle--large"
      >
        Dashboard
      </h1>
    </div>
    <div
      class="euiFlexItem euiFlexItem--flexGrowZero"
    >
      <a
        class="euiButton euiButton--primary euiButton--fill"
        data-test-subj="createChannelButton"
        href="#/channels"
        rel="noreferrer"
      >
        <span
          class="euiButtonContent euiButton__content"
        >
          <span
            class="euiButton__text"
          >
            Manage channels
          </span>
        </span>
      </a>
    </div>
  </div>
  <div
    class="euiSpacer euiSpacer--m"
  />
  <div
    class="euiFlexGroup euiFlexGroup--gutterSmall euiFlexGroup--alignItemsCenter euiFlexGroup--directionRow euiFlexGroup--responsive"
  >
    <div
      class="euiFlexItem"
    >
      <div
        class="euiFormControlLayout euiFormControlLayout--fullWidth"
      >
        <div
          class="euiFormControlLayout__childrenWrapper"
        >
          <input
            class="euiFieldSearch euiFieldSearch--fullWidth"
            data-test-subj="notifications-search-bar-input"
            placeholder="Search"
            type="search"
            value="test"
          />
          <div
            class="euiFormControlLayoutIcons"
          >
            <span
              class="euiFormControlLayoutCustomIcon"
            >
              <svg
                aria-hidden="true"
                class="euiIcon euiIcon--medium euiIcon-isLoading euiFormControlLayoutCustomIcon__icon"
                focusable="false"
                height="16"
                role="img"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              />
            </span>
          </div>
        </div>
      </div>
    </div>
    <div
      class="euiFlexItem euiFlexItem--flexGrowZero"
    >
      <div
        class="euiFlexGroup euiFlexGroup--gutterSmall euiFlexGroup--directionRow euiSuperDatePicker__flexWrapper euiSuperDatePicker__flexWrapper--noUpdateButton"
      >
        <div
          class="euiFlexItem"
        >
          <div
            class="euiFormControlLayout euiFormControlLayout--group euiSuperDatePicker"
          >
            <div
              class="euiPopover euiPopover--anchorDownLeft"
              id="QuickSelectPopover"
            >
              <div
                class="euiPopover__anchor euiQuickSelectPopover__anchor"
              >
                <button
                  aria-label="Date quick select"
                  class="euiButtonEmpty euiButtonEmpty--primary euiButtonEmpty--xSmall euiFormControlLayout__prepend"
                  data-test-subj="superDatePickerToggleQuickMenuButton"
                  type="button"
                >
                  <span
                    class="euiButtonContent euiButtonContent--iconRight euiButtonEmpty__content"
                  >
                    <svg
                      aria-hidden="true"
                      class="euiIcon euiIcon--medium euiIcon-isLoading euiButtonContent__icon"
                      focusable="false"
                      height="16"
                      role="img"
                      viewBox="0 0 16 16"
                      width="16"
                      xmlns="http://www.w3.org/2000/svg"
                    />
                    <span
                      class="euiButtonEmpty__text euiQuickSelectPopover__buttonText"
                    >
                      <svg
                        aria-hidden="true"
                        class="euiIcon euiIcon--medium euiIcon-isLoading"
                        focusable="false"
                        height="16"
                        role="img"
                        viewBox="0 0 16 16"
                        width="16"
                        xmlns="http://www.w3.org/2000/svg"
                      />
                    </span>
                  </span>
                </button>
              </div>
            </div>
            <div
              class="euiFormControlLayout__childrenWrapper"
            >
              <div
                class="euiDatePickerRange euiDatePickerRange--inGroup"
              >
                <button
                  class="euiSuperDatePicker__prettyFormat"
                  data-test-subj="superDatePickerShowDatesButton"
                >
                  Last 7 days
                  <span
                    class="euiSuperDatePicker__prettyFormatLink"
                  >
                    Show dates
                  </span>
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="euiFlexItem euiFlexItem--flexGrowZero"
    >
      <button
        class="euiButton euiButton--primary"
        type="button"
      >
        <span
          class="euiButtonContent euiButton__content"
        >
          <svg
            aria-hidden="true"
            class="euiIcon euiIcon--medium euiIcon-isLoading euiButtonContent__icon"
            focusable="false"
            height="16"
            role="img"
            viewBox="0 0 16 16"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          />
          <span
            class="euiButton__text"
          >
            Refresh
          </span>
        </span>
      </button>
    </div>
  </div>
  <div
    class="euiSpacer euiSpacer--s"
  />
  <div
    class="euiFlexGroup euiFlexGroup--gutterExtraSmall euiFlexGroup--alignItemsCenter euiFlexGroup--directionRow euiFlexGroup--wrap"
    style="min-height: 32px;"
  >
    <div
      class="euiFlexItem euiFlexItem--flexGrowZero"
    >
      <div
        class="euiPopover euiPopover--anchorRightUp euiPopover--withTitle"
      >
        <div
          class="euiPopover__anchor"
        >
          <button
            aria-label="Change all filters"
            class="euiButtonIcon euiButtonIcon--primary"
            title="Change all filters"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="euiIcon euiIcon--medium euiIcon-isLoading euiButtonIcon__icon"
              focusable="false"
              height="16"
              role="img"
              viewBox="0 0 16 16"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            />
          </button>
        </div>
      </div>
    </div>
    <div
      class="euiFlexItem euiFlexItem--flexGrowZero"
    >
      <div
        class="euiPopover euiPopover--anchorDownLeft euiPopover--withTitle"
      >
        <div
          class="euiPopover__anchor"
        >
          <button
            class="euiButtonEmpty euiButtonEmpty--primary euiButtonEmpty--xSmall"
            type="button"
          >
            <span
              class="euiButtonContent euiButtonEmpty__content"
            >
              <span
                class="euiButtonEmpty__text"
              >
                + Add filter
              </span>
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    class="euiSpacer euiSpacer--l"
  />
  <div
    class="euiPanel euiPanel--paddingMedium"
  >
    <div
      class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
      style="padding: 0px 0px;"
    >
      <div
        class="euiFlexItem"
      >
        <h3
          class="euiTitle euiTitle--medium"
        >
          Notification history
        </h3>
      </div>
    </div>
    <hr
      class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginSmall"
    />
    <div
      style="padding: initial;"
    >
      <div
        class="euiBasicTable euiBasicTable-loading"
      >
        <div>
          <div
            class="euiTableHeaderMobile"
          >
            <div
              class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsBaseline euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow"
            >
              <div
                class="euiFlexItem euiFlexItem--flexGrowZero"
              />
              <div
                class="euiFlexItem euiFlexItem--flexGrowZero"
              >
                <div
                  class="euiTableSortMobile"
                >
                  <div
                    class="euiPopover euiPopover--anchorDownRight"
                  >
                    <div
                      class="euiPopover__anchor"
                    >
                      <button
                        class="euiButtonEmpty euiButtonEmpty--primary euiButtonEmpty--xSmall euiButtonEmpty--flushRight"
                        type="button"
                      >
                        <span
                          class="euiButtonContent euiButtonContent--iconRight euiButtonEmpty__content"
                        >
                          <svg
                            aria-hidden="true"
                            class="euiIcon euiIcon--medium euiIcon-isLoading euiButtonContent__icon"
                            focusable="false"
                            height="16"
                            role="img"
                            viewBox="0 0 16 16"
                            width="16"
                            xmlns="http://www.w3.org/2000/svg"
                          />
                          <span
                            class="euiButtonEmpty__text"
                          >
                            Sorting
                          </span>
                        </span>
                      </button>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <table
            class="euiTable euiTable--responsive euiTable--auto"
            id="random_html_id"
            tabindex="-1"
          >
            <caption
              class="euiScreenReaderOnly euiTableCaption"
            />
            <thead>
              <tr>
                <th
                  aria-live="polite"
                  aria-sort="none"
                  class="euiTableHeaderCell"
                  data-test-subj="tableHeaderCell_event_source.title_0"
                  role="columnheader"
                  scope="col"
                >
                  <button
                    class="euiTableHeaderButton"
                    data-test-subj="tableHeaderSortButton"
                    type="button"
                  >
                    <span
                      class="euiTableCellContent"
                    >
                      <span
                        class="euiTableCellContent__text"
                        title="Notification"
                      >
                        Notification
                      </span>
                      <span
                        class="euiScreenReaderOnly"
                      >
                        Click to sort in ascending order
                      </span>
                    </span>
                  </button>
                </th>
                <th
                  aria-live="polite"
                  aria-sort="none"
                  class="euiTableHeaderCell"
                  data-test-subj="tableHeaderCell_event_source.feature_1"
                  role="columnheader"
                  scope="col"
                >
                  <button
                    class="euiTableHeaderButton"
                    data-test-subj="tableHeaderSortButton"
                    type="button"
                  >
                    <span
                      class="euiTableCellContent"
                    >
                      <span
                        class="euiTableCellContent__text"
                        title="Source type"
                      >
                        Source type
                      </span>
                      <span
                        class="euiScreenReaderOnly"
                      >
                        Click to sort in ascending order
                      </span>
                    </span>
                  </button>
                </th>
                <th
                  aria-live="polite"
                  aria-sort="none"
                  class="euiTableHeaderCell"
                  data-test-subj="tableHeaderCell_event_source.severity_2"
                  role="columnheader"
                  scope="col"
                >
                  <button
                    class="euiTableHeaderButton"
                    data-test-subj="tableHeaderSortButton"
                    type="button"
                  >
                    <span
                      class="euiTableCellContent"
                    >
                      <span
                        class="euiTableCellContent__text"
                        title="Severity"
                      >
                        Severity
                      </span>
                      <span
                        class="euiScreenReaderOnly"
                      >
                        Click to sort in ascending order
                      </span>
                    </span>
                  </button>
                </th>
                <th
                  aria-live="polite"
                  aria-sort="descending"
                  class="euiTableHeaderCell"
                  data-test-subj="tableHeaderCell_last_updated_time_ms_3"
                  role="columnheader"
                  scope="col"
                >
                  <button
                    class="euiTableHeaderButton euiTableHeaderButton-isSorted"
                    data-test-subj="tableHeaderSortButton"
                    type="button"
                  >
                    <span
                      class="euiTableCellContent"
                    >
                      <span
                        class="euiTableCellContent__text"
                        title="Time sent; Sorted in descending order"
                      >
                        Time sent
                      </span>
                      <svg
                        aria-hidden="true"
                        aria-label="Sorted in descending order"
                        class="euiIcon euiIcon--medium euiIcon-isLoading euiTableSortIcon"
                        focusable="false"
                        height="16"
                        role="img"
                        viewBox="0 0 16 16"
                        width="16"
                        xmlns="http://www.w3.org/2000/svg"
                      />
                      <span
                        class="euiScreenReaderOnly"
                      >
                        Click to sort in ascending order
                      </span>
                    </span>
                  </button>
                </th>
                <th
                  class="euiTableHeaderCell"
                  data-test-subj="tableHeaderCell_success_4"
                  role="columnheader"
                  scope="col"
                >
                  <div
                    class="euiTableCellContent"
                  >
                    <span
                      class="euiTableCellContent__text"
                      title="Sent status"
                    >
                      Sent status
                    </span>
                  </div>
                </th>
                <th
                  class="euiTableHeaderCell"
                  data-test-subj="tableHeaderCell_status_list_5"
                  role="columnheader"
                  scope="col"
                >
                  <div
                    class="euiTableCellContent"
                  >
                    <span
                      class="euiTableCellContent__text"
                      title="Channels"
                    >
                      Channels
                    </span>
                  </div>
                </th>
                <th
                  class="euiTableHeaderCell"
                  data-test-subj="tableHeaderCell_status_list_6"
                  role="columnheader"
                  scope="col"
                >
                  <div
                    class="euiTableCellContent"
                  >
                    <span
                      class="euiTableCellContent__text"
                      title="Channel types"
                    >
                      Channel types
                    </span>
                  </div>
                </th>
              </tr>
            </thead>
            <tbody>
              <tr
                class="euiTableRow"
              >
                <td
                  class="euiTableRowCell euiTableRowCell--isMobileFullWidth"
                  colspan="7"
                >
                  <div
                    class="euiTableCellContent euiTableCellContent--alignCenter"
                  >
                    <span
                      class="euiTableCellContent__text"
                    >
                      No items found
                    </span>
                  </div>
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </div>
</div>
`;
